dbms/src/Storages/DeltaMerge/File/ColumnStream.cpp

```diff
@@ -0,0 +1,316 @@
+    const String & file_name_base;
+    ReadLimiterPtr read_limiter;
+
+private:
+    MarksInCompressedFilePtr loadRawColMarkTo(const MarksInCompressedFilePtr & res, size_t bytes_size)
+    {
+        auto file = reader.file_provider->newRandomAccessFile(
+            reader.dmfile->colMarkPath(file_name_base),
+            reader.dmfile->encryptionMarkPath(file_name_base));
+        PageUtil::readFile(file, 0, reinterpret_cast<char *>(res->data()), bytes_size, read_limiter);
+        return res;
+    }
+    MarksInCompressedFilePtr loadColMarkWithChecksum(const MarksInCompressedFilePtr & res, size_t bytes_size)
+    {
+        auto buffer = ChecksumReadBufferBuilder::build(
+            reader.file_provider,
+            reader.dmfile->colMarkPath(file_name_base),
+            reader.dmfile->encryptionMarkPath(file_name_base),
+            reader.dmfile->getConfiguration()->getChecksumFrameLength(),
+            read_limiter,
+            reader.dmfile->getConfiguration()->getChecksumAlgorithm(),
+            reader.dmfile->getConfiguration()->getChecksumFrameLength());
+        buffer->readBig(reinterpret_cast<char *>(res->data()), bytes_size);
+        return res;
+    }
+    MarksInCompressedFilePtr loadColMarkFromMetav2(const MarksInCompressedFilePtr & res, size_t bytes_size)
+    {
+        auto info = reader.dmfile->merged_sub_file_infos.find(reader.dmfile->colMarkFileName(file_name_base));
+        if (info == reader.dmfile->merged_sub_file_infos.end())
+        {
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Unknown mark file {}",
+                reader.dmfile->colMarkFileName(file_name_base));
+        }
+
+        auto file_path = reader.dmfile->mergedPath(info->second.number);
+        auto encryp_path = reader.dmfile->encryptionMergedPath(info->second.number);
+        auto offset = info->second.offset;
+        auto data_size = info->second.size;
+
+        if (data_size == 0)
+            return res;
+
+        // first read from merged file to get the raw data(contains the header)
+        auto buffer = ReadBufferFromRandomAccessFileBuilder::build(
+            reader.file_provider,
+            file_path,
+            encryp_path,
+            reader.dmfile->getConfiguration()->getChecksumFrameLength(),
+            read_limiter);
+        buffer.seek(offset);
+
+        String raw_data;
+        raw_data.resize(data_size);
+
+        buffer.read(reinterpret_cast<char *>(raw_data.data()), data_size);
+        // read from the buffer based on the raw data
+        auto buf = ChecksumReadBufferBuilder::build(
+            std::move(raw_data),
+            reader.dmfile->colDataPath(file_name_base),
+            reader.dmfile->getConfiguration()->getChecksumFrameLength(),
+            reader.dmfile->configuration->getChecksumAlgorithm(),
+            reader.dmfile->configuration->getChecksumFrameLength());
+        buf->readBig(reinterpret_cast<char *>(res->data()), bytes_size);
```

Comment:
`FramedChecksumReadBuffer` also support `seek`, why do we need a `MemoryRandomAccessFile`?

Reply:
The `merge.x` file is created by using `ReadBufferFromRandomAccessFile` to copy the raw data from other files that are written by `FramedChecksumWriteBuffer`. So the layout of `merge.x` file is 
```
| <---                       file0                     ---> | <---          file1          ---> |
| frame0_header | frame0_data | frame1_header | frame1_data | frame0_header | frame0_data | ... |
                                                            ^
                                                   offset_in_merged_file
```

The `offset_in_merged_file` contains the size of `frame header` and real `data`.

If we open the whole `merge.x` as a `FramedChecksumWriteBuffer`, it will try to read `offset_in_merged_file` from the `frame_data`, without counting the size of `frame_header`. That's not the offset we want.

https://github.com/pingcap/tiflash/blob/c2b76f9214e1cad9388ed10bb43d9d26e8934098/dbms/src/Storages/DeltaMerge/File/DMFile.cpp#L1174-L1195

Reply:
Maybe can `FramedChecksumReadBuffer::seek(offset_in_merged_file , SEEK_CUR)` can work?